Quantum: Add initial qltests for OpenSSL modeling

bdrodes opened the pull request.

jketema commented:
Just a quick drive-by comment: the tests should preferably go into the `tests/experimental` sub-directory to match what is being done for the library itself.

nicolaswill commented:
> Just a quick drive-by comment: the tests should preferably go into the `tests/experimental/library-tests` sub-directory to match what is being done for the library itself.

@jketema Would `tests/library-tests/experimental/quantum/*` also work? I do not have any preference, but we will need to update CODEOWNERS to specify whatever directory the tests eventually live in. If we use the structure I suggest, the existing CODEOWNERS entries should cover those test directories already.

jketema commented:
@nicolaswill How about changing the pattern in the code owners file to `**/experimental/**/quantum/ @github/ps-codeql`. I believe that should work?

nicolaswill commented on cpp/ql/lib/experimental/quantum/OpenSSL/CtxFlow.qll:
```diff
@@ -29,7 +29,19 @@ import semmle.code.cpp.dataflow.new.DataFlow
+  CtxType() {
+    // It is possible for users to use the underlying type of the CTX variables
+    // these have a name matching 'evp_%ctx_%st
+    this.getUnspecifiedType().stripType().getName().matches("evp_%ctx_%st")
```
For a C team reviewier, I am curious about necessity of the logic in this charpred (see comment about "build mode none issues" on the lines below.

jketema replied:
If this is in fact only needed for your test to work, then I would strongly suggest to rewrite the test.

jketema reviewed (dismissed):
This is starting to pull in significant parts of what looks to be derived from the OpenSSL headers into our test directories. OpenSSL has a different license from ours (Apache vs. MIT). Is this legally ok?

nicolaswill commented:
> This is starting to pull in significant parts of what looks to be derived from the OpenSSL headers into our test directories. OpenSSL has a different license from ours (Apache vs. MIT). Is this legally ok?

@jketema We stub out other libraries with the Apache license. An example is Spring.
@bdrodes We do, however, need OpenSSL's `LICENSE.txt` for the stubs (see the Spring example [here](https://github.com/github/codeql/tree/main/java/ql/test/stubs/springframework-5.8.x)).

jketema commented:
@nicolaswill Thanks for pointing our the prior art in the Java QL tests. I think we want a similar kind of set up where the stubs are isolated in their own directory, we have a top-level readme like https://github.com/github/codeql/blob/main/java/ql/test/stubs/README.md, and have a license file in the directory with the OpenSSL stubs.

bdrodes commented:
I added a stubs location and moved the openssl stubs accordingly, and added the openssl apache license.

bdrodes commented:
@jketema and @nicolaswill, updated the stubs one more time to put them under 'crypto' and added our alias to the code owners for any test/stubs/crypto moving forward.

jketema commented:
I think we want the stubs directory to also be under experimental. Note that Java has something similar. There's both a stubs directory under test and under test/experimental.

Could you also please fix the Ql-for-QL warnings?

nicolaswill reviewed (changes requested):
Comments above.
